crates/biome_js_analyze/src/lint/nursery/use_unified_type_signature.rs

```diff
@@ -0,0 +1,1221 @@
+            .syntax()
+            .first_leading_trivia()
+            .filter(|trivia| trivia.pieces().any(|piece| piece.is_comments()))
+            .is_some()
+        {
+            // Transfer comments and whitespace before removing the signature.
+            mutation.remove_node_and_transfer_trivia(
+                signature_to_remove_wrapper,
+                signature_to_extend_wrapper,
+            );
+        } else {
+            // Nothing important to transfer, just remove the signature.
+            mutation.remove_node(signature_to_remove_wrapper);
+        }
+
+        // As we go through the operations list, we collect what changes we need to make
+        // in order to provide the user with better description of the changes.
+        for parameter_info in &state.parameters_to_merge {
+            match &parameter_info.operation {
+                MergeParameterOperation::CombineTypeWithParameter(combine_with_param) => {
+                    let parameter_type = parameter_info
+                        .parameter
+                        .type_annotation()
+                        .and_then(|type_annotation| type_annotation.ty().ok())?;
+                    let combine_with_type = combine_with_param
+                        .type_annotation()
+                        .and_then(|type_annotation| type_annotation.ty().ok())?;
+                    mutation.replace_node(
+                        parameter_type.clone(),
+                        AnyTsType::TsUnionType(
+                            make::ts_union_type(make::ts_union_type_variant_list(
+                                [parameter_type, combine_with_type],
+                                [make::token(T![|])
+                                    .with_leading_trivia([(TriviaPieceKind::Whitespace, " ")])
+                                    .with_trailing_trivia([(TriviaPieceKind::Whitespace, " ")])],
+                            ))
+                            .build(),
+                        ),
+                    );
+                }
+                MergeParameterOperation::MakeOptional => {
+                    if let Some(optional_parameter) = parameter_info.parameter.as_optional() {
+                        mutation.replace_node(parameter_info.parameter.clone(), optional_parameter)
+                    }
+                }
+            }
+        }
+
+        let first_parameter_operation = state
+            .parameters_to_merge
+            .first()
+            .map(|info| &info.operation);
+        Some(JsRuleAction::new(
+            ctx.metadata().action_category(ctx.category(), ctx.group()),
+            ctx.metadata().applicability(),
+            markup! {
+                {
+                    if matches!(
+                        first_parameter_operation,
+                        Some(MergeParameterOperation::CombineTypeWithParameter(_))
+                    ) {
+                        "Combine overloads by using type union."
```

Comment:
Really nice work on the diagnostics and their suggestions!

Nit:

```suggestion
                        "Combine overloads using a type union."
```